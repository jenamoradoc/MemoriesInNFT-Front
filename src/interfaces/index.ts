export interface LoginResults {
  id: number;
  avatar_url: string;
  login: string;
  type: string;
}

export interface FetchResponse {
  items: LoginResults[];
}
