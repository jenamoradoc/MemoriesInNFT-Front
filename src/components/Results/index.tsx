import { FC, useEffect, useState } from "react";

import { LoginResults } from "../../interfaces";
import { getDataFn } from "../../utils";
import team from "../../assets/team.svg";
import "./results.css";

interface Props {
  login: string;
  results: LoginResults[];
  getResults: (data: LoginResults[]) => void;
  changeLoading: (state: boolean) => void;
  changeMessage: (message: string) => void;
}

interface SelectOption {
  value: string;
  text: string;
}

const Results: FC<Props> = ({
  login,
  results,
  getResults,
  changeLoading,
  changeMessage,
}) => {
  const [page, setPage] = useState<number>(1);

  console.log(results);

  const options: SelectOption[] = [
    { value: "best match", text: "Best Match" },
    { value: "followers", text: "Followers" },
    { value: "repositories", text: "Repositories" },
    { value: "joined", text: "Joined" },
  ];

  const [selected, setSelected] = useState(options[0].value);

  useEffect(() => {
    if (page > 1 && selected.trim() !== "") {
      getDataFn({
        login,
        page,
        setLoading: changeLoading,
        getResults,
        setMessage: changeMessage,
        sort: selected,
      });
    } else if (page === 1 && login.trim() !== "" && selected.trim() !== "") {
      getDataFn({
        login,
        setLoading: changeLoading,
        getResults,
        setMessage: changeMessage,
        sort: selected,
      });
    }

    // eslint-disable-next-line
  }, [page, selected]);

  const pages: number[] =
    results.length === 9 ? [1, 2, 3, 4, 5, 6, 7, 8, 9] : [1];

  return (
    <div>
      {results.length > 0 ? (
        <>
          <select
            className="results__sort"
            value={selected}
            onChange={({ target }) => setSelected(target.value)}
          >
            <option value="" disabled>
              --Choose an option
            </option>
            {options.map(({ value, text }) => (
              <option key={value} value={value}>
                {text}
              </option>
            ))}
          </select>
          <ul className="results__list">
            {results.map(({ id, login, avatar_url, type }) => (
              <li key={id} className="results__list-item">
                <img
                  src={avatar_url}
                  alt={login}
                  className="results__list-item-image"
                  draggable={false}
                />
                <strong className="results__list-item-login">{login}</strong>
                <span className="results__list-item-type">{type}</span>
              </li>
            ))}
          </ul>
          <div className="buttons__container">
            <button
              onClick={() => setPage((page) => page - 1)}
              disabled={page <= 1}
              style={{ borderRadius: "5px 0 0 5px" }}
            >
              &laquo;
            </button>
            {pages.map((pageNumber) => (
              <button
                key={pageNumber}
                onClick={() => setPage(pageNumber)}
                disabled={page === pageNumber}
                data-testid={pageNumber}
              >
                {pageNumber}
              </button>
            ))}
            <button
              onClick={() => setPage((page) => page + 1)}
              disabled={page === pages.length}
              style={{ borderRadius: "0 5px 5px 0" }}
            >
              &raquo;
            </button>
          </div>
        </>
      ) : (
        <div className="results__message">
          {login.trim().length === 0 ? (
            <>
              <h2>GitHub User Search App</h2>
              <img src={team} alt="Team" draggable={false} />
            </>
          ) : results.length === 0 ? (
            <>
              <h2>Results not found!</h2>
              <img src={team} alt="Team" draggable={false} />
            </>
          ) : null}
        </div>
      )}
    </div>
  );
};

export default Results;
