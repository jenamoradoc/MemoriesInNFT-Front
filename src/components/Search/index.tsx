import { FC, FormEvent, useState, useEffect } from "react";

import Spinner from "../Spinner";

import { LoginResults } from "../../interfaces";
import { getDataFn } from "../../utils";
import "./search.css";

interface Props {
  usernameEntered: (login: string) => void;
  getResults: (data: LoginResults[]) => void;
  changeLoading: (state: boolean) => void;
  loading: boolean;
  results: LoginResults[];
  changeMessage: (message: string) => void;
  message: string;
}

const Search: FC<Props> = ({
  usernameEntered,
  getResults,
  loading,
  changeLoading,
  results,
  changeMessage,
  message,
}) => {
  const [login, setLogin] = useState<string>("");

  useEffect(() => {
    if (results.length > 0) {
      getResults(results);
    }

    // eslint-disable-next-line
  }, [results]);

  const handleSubmit = async (
    event: FormEvent<HTMLFormElement>
  ): Promise<void> => {
    event.preventDefault();

    if (login.trim().length === 0) return;

    usernameEntered(login);
    getDataFn({
      login,
      setLoading: changeLoading,
      getResults,
      setMessage: changeMessage,
    });
  };

  const handleClearFilter = () => {
    if (login.trim().length > 0) {
      setLogin("");
      usernameEntered("");
      getResults([]);
    }
  };

  return (
    <>
      <aside className="search__form-container">
        <form onSubmit={handleSubmit} className="search__form">
          <h1 className="search__form-title">Search github user(s)</h1>
          <input
            type="text"
            value={login}
            autoFocus
            onChange={({ target }) => setLogin(target.value)}
            className="search__form-input"
          />

          <div className="search__form-btns__container">
            <button type="submit" className="search__form-btn">
              Submit
            </button>

            <button className="search__form-btn" onClick={handleClearFilter}>
              Clear Filter
            </button>
          </div>

          {message.trim().length > 0 && (
            <p className="error__message">{message}</p>
          )}
        </form>
      </aside>
      {loading && <Spinner />}
    </>
  );
};

export default Search;
