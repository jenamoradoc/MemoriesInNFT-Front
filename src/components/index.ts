import Results from "./Results";
import Search from "./Search";
import Spinner from "./Spinner";

export { Results, Search, Spinner };
