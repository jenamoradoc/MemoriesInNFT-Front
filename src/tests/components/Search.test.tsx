import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

import { Search } from "../../components";
import { LoginResults } from "../../interfaces";

describe("Testing <Search />", () => {
  const changeLoading = jest.fn();
  const changeMessage = jest.fn();
  const getResults = jest.fn();
  const getLoginQuery = jest.fn();
  const loading: boolean = false;
  const message: string = "";
  const results: LoginResults[] = [];

  test("should show the initial screen", () => {
    const { container } = render(
      <Search
        changeLoading={changeLoading}
        changeMessage={changeMessage}
        getResults={getResults}
        usernameEntered={getLoginQuery}
        loading={loading}
        message={message}
        results={results}
      />
    );

    expect(container).toMatchSnapshot();
  });

  test("should focus the input field", () => {
    render(
      <Search
        changeLoading={changeLoading}
        changeMessage={changeMessage}
        getResults={getResults}
        usernameEntered={getLoginQuery}
        loading={loading}
        message={message}
        results={results}
      />
    );

    expect(screen.getByRole("textbox")).toHaveFocus();
  });

  test("should allow typing in the text field", () => {
    render(
      <Search
        changeLoading={changeLoading}
        changeMessage={changeMessage}
        getResults={getResults}
        usernameEntered={getLoginQuery}
        loading={loading}
        message={message}
        results={results}
      />
    );

    userEvent.type(screen.getByRole("textbox"), "pablo");
    userEvent.click(screen.getByText(/submit/i));
    expect((screen.getByRole("textbox") as HTMLInputElement).value).toBe(
      "pablo"
    );
  });
});
