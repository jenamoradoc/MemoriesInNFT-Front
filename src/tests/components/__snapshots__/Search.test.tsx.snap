// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing <Search /> should show the initial screen 1`] = `
<div>
  <aside
    class="search__form-container"
  >
    <form
      class="search__form"
    >
      <h1
        class="search__form-title"
      >
        Search github user(s)
      </h1>
      <input
        class="search__form-input"
        type="text"
        value=""
      />
      <div
        class="search__form-btns__container"
      >
        <button
          class="search__form-btn"
          type="submit"
        >
          Submit
        </button>
        <button
          class="search__form-btn"
        >
          Clear Filter
        </button>
      </div>
    </form>
  </aside>
</div>
`;
